imap/sievedir.c

```diff
@@ -0,0 +1,426 @@
+        struct stat sbuf;
+
+        if (!strcmp(name, ".") || !strcmp(name, "..")) continue;
+
+        strlcpy(path + dir_len, name, sizeof(path) - dir_len);
+
+        if (lstat(path, &sbuf) < 0) continue;
+
+        if (S_ISLNK(sbuf.st_mode)) {
+            /* fetch link target */
+            ssize_t tgt_len = readlink(path, target, sizeof(target) - 1);
+
+            if (tgt_len > 0) target[tgt_len] = '\0';
+        }
+        else if (!S_ISREG(sbuf.st_mode)) {
+            /* ignore irregular files */
+            continue;
+        }
+
+        if (func(sievedir, name, &sbuf, target, rock) != 0) break;
+    }
+
+    closedir(dp);
+}
+
+struct count_rock {
+    int count;
+    const char *myname;
+};
+
+static int count_cb(const char *sievedir __attribute__((unused)),
+                    const char *name, struct stat *sbuf,
+                    const char *link_target __attribute__((unused)),
+                    void *rock)
+{
+    struct count_rock *crock = (struct count_rock *) rock;
+    size_t name_len;
+
+    if (!S_ISREG(sbuf->st_mode)) return 0;
+
+    name_len = strlen(name);
+    if (name_len > SCRIPT_SUFFIX_LEN &&
+        !strcmp(name + name_len - SCRIPT_SUFFIX_LEN, SCRIPT_SUFFIX)) {
+        /* is a script */
+        name_len -= SCRIPT_SUFFIX_LEN;
+        if (!crock->myname ||
+            strlen(crock->myname) != name_len ||
+            strncmp(crock->myname, name, name_len)) {
+            /* and it's different from me */
+            crock->count++;
+        }
+    }
+
+    return 0;
+}
+
+/* counts the number of scripts user has that are DIFFERENT from name */
+EXPORTED int sievedir_num_scripts(const char *sievedir, const char *name)
+{
+    struct count_rock crock = { 0, name };
+
+    sievedir_foreach(sievedir, &count_cb, &crock);
+
+    return crock.count;
+}
+
+
+EXPORTED struct buf *sievedir_get_script(const char *sievedir,
+                                         const char *script)
+{
+    struct buf buf = BUF_INITIALIZER;
+
+    buf_printf(&buf, "%s/%s", sievedir, script);
+
+    int fd = open(buf_cstring(&buf), 0);
+    if (fd < 0) return NULL;
+
+    buf_free(&buf);
+    buf_refresh_mmap(&buf, 1, fd, script, MAP_UNKNOWN_LEN, "sieve");
+
+    close(fd);
+
+    struct buf *ret = buf_new();
+
+    buf_move(ret, &buf);
+
+    return ret;
+}
+ 
+/* Everything but '/' and '\0' are valid. */
+EXPORTED int sievedir_valid_name(const struct buf *name)
+{
+    size_t lup, len = buf_len(name);
+    const char *ptr;
+
+    /* must be at least one character long */
+    if (len < 1) return 0;
+
+    ptr = buf_base(name);
+
+    for (lup = 0; lup < len; lup++) {
+        if ((ptr[lup] == '/') || (ptr[lup] == '\0')) return 0;
+    }
+
+    return (lup < 1013);
+}
+
+EXPORTED int sievedir_script_exists(const char *sievedir, const char *name)
+{
+    char path[PATH_MAX];
+    struct stat sbuf;
+
+    snprintf(path, sizeof(path), "%s/%s%s", sievedir, name, SCRIPT_SUFFIX);
+
+    return (stat(path, &sbuf) == 0);
```

Comment:
In other places we ignore files that are not regular files, do we need to do the same here?  I guess the question is what this API reports.  Does it report "something with this name exists", or does it report "a usable script with this name exists"?

Reply:
https://github.com/cyrusimap/cyrus-imapd/commit/5044b936283aa8000a55f23077ca75a3d79d474a